[C++ API Parity] NLLLoss

PyExtreme opened the pull request:
Hi @yf225 , this is my initial commit on NLLLoss referenced here #25883 
There are few errors which I have just spotted. Will fix them and let you know.

jhaankit373 committed:
Add C++ Frontend - NLLLoss

jhaankit373 committed:
Add C++ Frontend - NLLLoss

yf225 commented on torch/csrc/api/include/torch/nn/options/loss.h:
```diff
@@ -153,5 +153,20 @@ struct TORCH_API TripletMarginLossOptions {
+  /// ``True``, the loss is averaged over non-ignored targets.
+  TORCH_ARG(double, ignore_index) = -100;
+  /// Specifies the reduction to apply to the output. Default: Mean
+  TORCH_ARG(Reduction::Reduction, reduction) = Reduction::Mean;
```
I think `Reduction` is now under `torch::`,  and we can write
```suggestion
  TORCH_ARG(torch::Reduction::Reduction, reduction) = torch::Reduction::Mean;
```

yf225 commented on torch/csrc/api/include/torch/nn/functional/loss.h:
```diff
@@ -127,6 +127,18 @@ inline Tensor triplet_margin_loss(
+      target,
+      options.weight(),
+      options.ignore_index(),
+      options.reduction());
```
It would be awesome to implement the equivalent of the Python version:
https://github.com/pytorch/pytorch/blob/783c9c84450f9b75d8cb9752062e961b7449e2c6/torch/nn/functional.py#L1830-L1859
to match the Python version even better :D Thanks so much for your help!

yf225 replied:
We can look at changes to `kl_div`/`mse_loss`/`binary_cross_entropy` as examples in https://github.com/pytorch/pytorch/pull/27942/files

PyExtreme replied:
Hi @yf225 , would be great if you could provide me some resources which I can study so as to easily translate the code part. I also had a look at the link posted, however, I am not sure about the alternative of `.view or .contiguous()` in C++. It really help me in my future works too.

yf225 commented on torch/csrc/api/src/nn/modules/loss.cpp:
```diff
@@ -201,5 +201,27 @@ Tensor SoftMarginLossImpl::forward(const Tensor& input, const Tensor& target) {
+}
+
+void NLLLossImpl::reset() {
+  register_buffer("weight", options.weight());
```
I think we can write
```suggestion
  weight = register_buffer("weight", options.weight());
```

yf225 commented on test/cpp/api/modules.cpp:
```diff
@@ -1868,6 +1884,12 @@ TEST_F(ModulesTest, PrettyPrintSoftMarginLoss) {
+TEST_F(ModulesTest, PrettyPrintLoss) {
+  ASSERT_EQ(
+      c10::str(NLLLoss(NLLLossOptions().ignore_index(-100))),
+      "torch::nn::NLLLoss(ignore_index=-100)");
```
I believe in Python version it doesn't print any options arguments, and we can simplify this to just check for `"torch::nn::NLLLoss()"` here :D

yf225 reviewed (commented):
@PyExtreme Thanks so much for the awesome work! I left some comments :D

PyExtreme commented:
Thanks @yf225  for the review. Will make the changes and make a fresh PR.

PyExtreme commented:
Hi @yf225 , I have been reading, understanding the codebase and playing with the C++ code. I have made some changes and a few are left to be done. I maybe on vacation till next complete week. So, I may be pushing changes next to next week.

Thanks

jhaankit373 committed:
Add C++ Parity - NLLLoss

jhaankit373 committed:
C++ API Parity: NLLLoss

yf225 commented:
@PyExtreme Thanks a lot for your help! Please keep me posted :D

jhaankit373 committed:
c++ api parity: nllloss

jhaankit373 committed:
cpp api frontend: nllloss

PyExtreme commented:
Hi @yf225 , I have compiled and run test which passed.
Would be great if you could please review it once.

Thank you

PyExtreme commented:
@yf225 , would be great if you could please have a look at this commit.

I will fix the conflicts later.

Thank you

yf225 commented:
@PyExtreme Thanks so much for the update! I fixed the conflicts and I will review it tomorrow :D

PyExtreme commented:
Hi @yf225 , I have finished work for cross_entropy_loss also. However, when I am building it, I am getting error related to onnx while building which is causing it to fail. 

The command which I have executed is`
MAX_JOBS=8 DEBUG=1 USE_DISTRIBUTED=0 USE_MKLDNN=0 USE_CUDA=0 BUILD_TEST=1 USE_FBGEMM=0 USE_NNPACK=0 USE_QNNPACK=0 python setup.py develop`

The error which I am getting is:

```
../caffe2/opt/onnxifi_op.cc:95:7: error: ‘onnxTensorDescriptorV1 {aka struct onnxTensorDescriptorV1}’ has no member named ‘isOffline’
   to->isOffline = from->isOffline;
       ^~~~~~~~~
../caffe2/opt/onnxifi_op.cc: In function ‘void caffe2::{anonymous}::BlobToTensorDescriptor(const string&, caffe2::Workspace*, onnxTensorDescriptorV1*, std::vector<std::vector<long unsigned int, std::allocator<long unsigned int> > >*, std::vector<std::vector<float> >*, std::vector<std::vector<int> >*)’:
../caffe2/opt/onnxifi_op.cc:134:9: error: ‘onnxTensorDescriptorV1 {aka struct onnxTensorDescriptorV1}’ has no member named ‘isOffline’
   desc->isOffline = false;
```

yf225 commented:
> Hi @yf225 , I have finished work for cross_entropy_loss also. However, when I am building it, I am getting error related to onnx while building which is causing it to fail.
> 
> The command which I have executed is` MAX_JOBS=8 DEBUG=1 USE_DISTRIBUTED=0 USE_MKLDNN=0 USE_CUDA=0 BUILD_TEST=1 USE_FBGEMM=0 USE_NNPACK=0 USE_QNNPACK=0 python setup.py develop`
> 
> The error which I am getting is:
> 
> ```
> ../caffe2/opt/onnxifi_op.cc:95:7: error: ‘onnxTensorDescriptorV1 {aka struct onnxTensorDescriptorV1}’ has no member named ‘isOffline’
>    to->isOffline = from->isOffline;
>        ^~~~~~~~~
> ../caffe2/opt/onnxifi_op.cc: In function ‘void caffe2::{anonymous}::BlobToTensorDescriptor(const string&, caffe2::Workspace*, onnxTensorDescriptorV1*, std::vector<std::vector<long unsigned int, std::allocator<long unsigned int> > >*, std::vector<std::vector<float> >*, std::vector<std::vector<int> >*)’:
> ../caffe2/opt/onnxifi_op.cc:134:9: error: ‘onnxTensorDescriptorV1 {aka struct onnxTensorDescriptorV1}’ has no member named ‘isOffline’
>    desc->isOffline = false;
> ```

Would you like to also open a PR for cross_entropy_loss? I can review/iterate on it and get it ready for our new version release, which is scheduled early next week. Thanks!

PyExtreme commented:
> > Hi @yf225 , I have finished work for cross_entropy_loss also. However, when I am building it, I am getting error related to onnx while building which is causing it to fail.
> > The command which I have executed is` MAX_JOBS=8 DEBUG=1 USE_DISTRIBUTED=0 USE_MKLDNN=0 USE_CUDA=0 BUILD_TEST=1 USE_FBGEMM=0 USE_NNPACK=0 USE_QNNPACK=0 python setup.py develop`
> > The error which I am getting is:
> > ```
> > ../caffe2/opt/onnxifi_op.cc:95:7: error: ‘onnxTensorDescriptorV1 {aka struct onnxTensorDescriptorV1}’ has no member named ‘isOffline’
> >    to->isOffline = from->isOffline;
> >        ^~~~~~~~~
> > ../caffe2/opt/onnxifi_op.cc: In function ‘void caffe2::{anonymous}::BlobToTensorDescriptor(const string&, caffe2::Workspace*, onnxTensorDescriptorV1*, std::vector<std::vector<long unsigned int, std::allocator<long unsigned int> > >*, std::vector<std::vector<float> >*, std::vector<std::vector<int> >*)’:
> > ../caffe2/opt/onnxifi_op.cc:134:9: error: ‘onnxTensorDescriptorV1 {aka struct onnxTensorDescriptorV1}’ has no member named ‘isOffline’
> >    desc->isOffline = false;
> > ```
> 
> Would you like to also open a PR for cross_entropy_loss? I can review/iterate on it and get it ready for our new version release, which is scheduled early next week. Thanks!
> 
> The build error is likely because of a submodule update problem, and it should go away if you do `git clean -xdf` and then `git submodule update`, and then build again.

Hi @yf225 , I had tried updating submodules but it didn't work. As suggested by @ptrblck , I have recloned and am rebuilding it. Once, I am done, I will make a new PR with both the losses. Expecting to complete it by EOD.

Thank you

PyExtreme commented:
@yf225 , I have added cross entropy loss and moved this to a new PR as I had to reclone the repo and build again.
Here is a link to it
https://github.com/pytorch/pytorch/pull/29812
